### ISSUE TITLE
Compilation issues with Phoenix Live Dashboard

### ISSUE THREAD

**Issue by alexcastano** [appsignal/appsignal-elixir#319]
If I add:

```
# config/config.exs
config :phoenix, :template_engines,
  eex: Appsignal.Phoenix.Template.EExEngine,
  exs: Appsignal.Phoenix.Template.ExsEngine
```

Then https://github.com/edgurgel/verk_web package breaks in compilation time:

```
== Compilation error in file web/views/shared_view.ex ==
** (UndefinedFunctionError) function Appsignal.Phoenix.Template.EExEngine.compile/2 is undefined (module Appsignal.Phoenix.Template.EExEngine is not available)
    Appsignal.Phoenix.Template.EExEngine.compile("web/templates/shared/job_modal.html.eex", "job_modal.html")
    (phoenix) lib/phoenix/template.ex:378: Phoenix.Template.compile/2
    (phoenix) lib/phoenix/template.ex:186: anonymous fn/3 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix) expanding macro: Phoenix.Template.__before_compile__/1
    web/views/shared_view.ex:1: VerkWeb.SharedView (module)
```

It seems it tries to use the `EExEngine` to compile its own templates but it fails because it is trying to use `Appsignal` ones.

**Comment by jeffkreeftmeijer** [comment]
Hey @alexcastano,

Thanks for letting us know about this, and sorry about this compilation issue. I [tried reproducing your issue](https://github.com/jeffkreeftmeijer/appsignal-phoenix-example/commit/3707772e4d2e9c313f2890ae9c9268905cb71b42) in my example app, but wasn't able to get the compilation error you describe. Could you compare the verk_web branch to your application, so we can find what I'm doing differently?

**Comment by alexcastano** [comment]
I cannot reproduce it in your project. It happens to mine but randomly. I cannot find a pattern, maybe the compilation order... but not sure :disappointed: 

I let you know if I find something more.

**Comment by tombruijn** [comment]
Hi @alexcastano,

Please let us know if this keeps occurring when. I'll close this issue for now. Let's reopen it when it happens again.

**Comment by gjaldon** [comment]
This issue still exists and is a flaky error. I just experienced it in our umbrella app and no one else in our team has experienced it. It looks like `AppSignal.phoenix?` fails sometimes and it likely has to do with build order of deps. When that fails, the `Appsignal.Phoenix.Template.EExEngine` and other modules that rely on `AppSignal.phoenix?/1` don't get compiled.

The fix is to do `mix deps.clean appsignal` and then redo `mix do deps.get, compile` until this doesn't fail. 

The fix for this may need to be done in Elixir/Mix.

**Comment by NobbZ** [comment]
As I have already mentioned elsewhere, no need to repeat a full `deps.clean`, just `mix do deps.clean appsignal, deps.build` **after** you made that `phoenix` is available.

In theory, as soon as `phoenix` is a dependency of your app, it should always be compiled before appsignal, due to the optional dependency.

If though you have a pristine environment and compile your dependencies in a sub-app of the umbrella that doesn't see the phoenix dependency, then appsignal will be compiled without the phoenix support. And it wont get recompiled for any of the other applications.

Thats why you need to clean its artifacts and manually compile it.

In my opinion, this is not a problem of `mix` but your project architecture/workflow. Just make sure you always build your deps from the umbrellas root.

Nevertheless, unless there is a hard compile time dependency on phoenix in those module, I'd just let them compile unconditionally.

**Comment by jeffkreeftmeijer** [comment]
Hey all, 

Thanks for chiming in, here, and thanks for the suggestions! We'd love to remove the optional compilation of the template engine modules, but we haven't found a way to do so, because they wrap the `Phoenix.Template.EexEngine` and `Phoenix.Template.ExsEngine` `compile/2` calls, respectively. If we remove the if-statement there, we'll fire a warning during compilation. Any suggestions here would be appreciated. :)

Also, we've looked into moving the Phoenix-specific code out of the main project and into a separate package which has a hard dependency on Phoenix. However, we don't want to make breaking changes like this, as we have some other changes coming up in the near future.

**Comment by odarriba** [comment]
Hi!

This is happening also with `oban_web` the web UI for [oban](https://github.com/sorentwo/oban). It is still closed source until it reaches open beta, but we are seeing same issue as commented here:

```
==> oban_web
Compiling 7 files (.ex)

== Compilation error in file lib/oban_web/views/dashboard_view.ex ==
** (UndefinedFunctionError) function Appsignal.Phoenix.Template.EExEngine.compile/2 is undefined (module Appsignal.Phoenix.Template.EExEngine is not available)
    Appsignal.Phoenix.Template.EExEngine.compile("lib/oban_web/templates/dashboard/_blitz.html.eex", "_blitz.html")
    (phoenix) lib/phoenix/template.ex:354: Phoenix.Template.compile/3
    (phoenix) lib/phoenix/template.ex:165: anonymous fn/4 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir) lib/enum.ex:1948: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix) expanding macro: Phoenix.Template.__before_compile__/1
    lib/oban_web/views/dashboard_view.ex:1: ObanWeb.DashboardView (module)
    (elixir) lib/kernel/parallel_compiler.ex:229: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/7
```

One way of avoiding this, is force compiling `appsignal` before executing any other command that fires a compilation:

```
$ mix deps.compile appsignal 
```

**Comment by ZombieHarvester** [comment]
I just caught similar error with `MIX_ENV=prod mix compile` after I added Phoenix LiveView 0.12.1 to my project. 

```elixir
== Compilation error in file lib/phoenix/live_dashboard/layout_view.ex ==
** (UndefinedFunctionError) function Appsignal.Phoenix.Template.EExEngine.compile/2 is undefined (module Appsignal.Phoenix.Template.EExEngine is not available)
    Appsignal.Phoenix.Template.EExEngine.compile("lib/phoenix/live_dashboard/templates/layout/dash.html.eex", "dash.html")
    (phoenix 1.5.1) lib/phoenix/template.ex:351: Phoenix.Template.compile/3
    (phoenix 1.5.1) lib/phoenix/template.ex:166: anonymous fn/4 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir 1.10.2) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix 1.5.1) expanding macro: Phoenix.Template.__before_compile__/1
    lib/phoenix/live_dashboard/layout_view.ex:1: Phoenix.LiveDashboard.LayoutView (module)
```

Removing the following lines helped
```elixir
# config/config.exs
config :phoenix, :template_engines,
  eex: Appsignal.Phoenix.Template.EExEngine,
  exs: Appsignal.Phoenix.Template.ExsEngine
```

**Comment by jeffkreeftmeijer** [comment]
Hey @ZombieHarvester,

Do you consistently get the error in your app? That might allow us to try and reproduce the problem. 

Removing the template engines from your configuration will definitely help, but it will remove template instrumentation from your samples in AppSignal. The rest of the integration will work as expected, though.

**Comment by ZombieHarvester** [comment]
Hey @jeffkreeftmeijer 

Yes I do, after upgrade to Phoenix 1.5. I had to remove template instrumentation because it simply couldn't compile

I have reproduced it in new app: https://github.com/ZombieHarvester/appsignal-issue

```elixir
> mix compile
...
==> phoenix_live_dashboard
Compiling 32 files (.ex)

== Compilation error in file lib/phoenix/live_dashboard/layout_view.ex ==
** (UndefinedFunctionError) function Appsignal.Phoenix.Template.EExEngine.compile/2 is undefined (module Appsignal.Phoenix.Template.EExEngine is not available)
    Appsignal.Phoenix.Template.EExEngine.compile("lib/phoenix/live_dashboard/templates/layout/dash.html.eex", "dash.html")
    (phoenix 1.5.1) lib/phoenix/template.ex:351: Phoenix.Template.compile/3
    (phoenix 1.5.1) lib/phoenix/template.ex:166: anonymous fn/4 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir 1.10.3) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix 1.5.1) expanding macro: Phoenix.Template.__before_compile__/1
    lib/phoenix/live_dashboard/layout_view.ex:1: Phoenix.LiveDashboard.LayoutView (module)
    (elixir 1.10.3) lib/kernel/parallel_compiler.ex:304: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/7
could not compile dependency :phoenix_live_dashboard, "mix compile" failed. You can recompile this dependency with "mix deps.compile phoenix_live_dashboard", update it with "mix deps.update phoenix_live_dashboard" or clean it with "mix deps.clean phoenix_live_dashboard"
```

**Comment by jeffkreeftmeijer** [comment]
Hey @ZombieHarvester,

Thanks a lot for explaining and setting up that example app. I've tried running your app, and found that since phoenix_live_dashboard uses templates. If we use `Phoenix.Template.EExEngine' as our template engine, phoenix_live_dashboard tries to use it to precompile templates. If phoenix_live_dashboard is compiled before AppSignal, it doesn't exist yet, and your app's compilation will fail.

I'm looking into nicer ways to resolve this issue in the future. For now, moving the AppSignal depenency above the phoenix_live_view one resolves this issue for me:

```
  defp deps do
    [
      {:appsignal, "~> 1.0"},
      {:phoenix, "~> 1.5.1"},
      {:phoenix_html, "~> 2.11"},
      {:phoenix_live_reload, "~> 1.2", only: :dev},
      {:phoenix_live_dashboard, "~> 0.2.0"},
      {:telemetry_metrics, "~> 0.4"},
      {:telemetry_poller, "~> 0.4"},
      {:gettext, "~> 0.11"},
      {:jason, "~> 1.0"},
      {:plug_cowboy, "~> 2.0"}
    ]
  end
```

**Comment by ZombieHarvester** [comment]
@jeffkreeftmeijer thanks, it seems to work. Hope there will be a more elegant solution in the future

**Comment by sebn** [comment]
> I'm looking into nicer ways to resolve this issue in the future.

https://github.com/phoenixframework/phoenix_live_dashboard/issues/60#issuecomment-614932788

In case it helps 🙂

**Comment by josevalim** [comment]
FWIW, compilation order is not guaranteed to follow `mix.exs` ordering, so moving app_signal up may fail in order situations. Ideally we would need to find a way for AppSignal to get the data they need without replacing the template engines.

**Comment by ZombieHarvester** [comment]
This needs to be an open issue, imho

**Comment by sebn** [comment]
I confirm compilation can still fail randomly even when `appsignal` is listed before `phoenix_live_dashboard` (reproduced just now).

**Comment by jeffkreeftmeijer** [comment]
Thanks for chiming in, all. We've been working on a new version of the Phoenix integration we're hoping to release soon. In the new version, we're moving off the template engines and added [the `Appsignal.View` module](
https://github.com/appsignal/appsignal-elixir-phoenix/blob/master/lib/appsignal_phoenix/view.ex
), which reimplements the `render/2` function to add instrumentation, which has been working great in our internal testing.

As for the current version, please let us know if you're still running into compilation issues when moving the appsignal dependency up. We'll look into backporting the new module if that's required.

**Comment by hubertlepicki** [comment]
@jeffkreeftmeijer to clarify, this new version of Phoenix integration is not in 13.3.3 just yet? I'm having the same issue at the very moment.

**Comment by jeffkreeftmeijer** [comment]
Hey Hubert,

That's right, the new template instrumentation hasn't been released in the current version of the integration. We're currently testing the new version of the integration (which includes the new view instrumentation module), and we should reach the first beta stage soon, when it's feature-complete.

We'd love to find a way to get this to work with the current version of the integration, but we haven't found a way to do so without breaking backwards compatibility yet. 

In the meantime, if you'd like to try the new version out, upgrade appsignal to 2.0.0-alpha.5 and let us know if you run into anything through support, or by opening an issue here. We'll post more detailed testing instructions when we reach the first beta.

You're right about this issue being closed while it shouldn't. Fixed!

**Comment by hubertlepicki** [comment]
Hey @jeffkreeftmeijer that's great. Unfortunately I don't think on this particular project adding an -alpha hex package will pass through code review, but I will attempt to try it on some other project.

The new instrumentation is based on Telemetry I assume?

**Comment by kenny-evitt** [comment]
@jeffkreeftmeijer It's not ideal, but [the suggestion by @odarriba](https://github.com/appsignal/appsignal-elixir/issues/319#issuecomment-560427670) to run `mix deps.compile appsignal --include-children` before `mix compile` seems to work for now.

**Comment by cblavier** [comment]
If you face the problem on Heroku, here is how you can apply @odarriba suggestion: add `hook_pre_compile="mix deps.compile appsignal --include-children"` to your `elixir_buildpack.config` file

**Comment by jeffkreeftmeijer** [comment]
Thanks for the suggestion, we'll recommend that for now. We're still working on that 2.0, which splits the project up into separate parts. We're in the beta stage now, and we're hoping to release that soon to finally squash this issue. :)

**Comment by jeffkreeftmeijer** [comment]
We've released version 2.0 of the AppSignal for Elixir integration, which should resolve this issue by introducing the `Appsignal.View` module I talked about [previously](https://github.com/appsignal/appsignal-elixir/issues/319#issuecomment-642702867). If you're running into this issue, please try 2.0 by upgrading using our [upgrade guide](https://docs.appsignal.com/elixir/installation/upgrading-from-1.x-to-2.x.html).

Feel free to reach out through [support](mailto: support@appsignal.com) if you run into any issues upgrading, or by opening a new issue here.

Thanks for your help getting to the bottom of this and finding workarounds for 1.x, everyone!
